### ISSUE TITLE
Non-termination in generic function type instantiation

### ISSUE THREAD

**Issue by kmillikin** [microsoft/pyright#3308]
I can trigger a stack overflow with the following code.  This is not real-world code.

```python
from random import random
from typing import TypeVar

T = TypeVar('T')
U = TypeVar('U')

def f(x: T, y: U) -> T | U:
  return x if random() > 0.5 else y

def g(x: T, y: T) -> T:
  return x if random() > 0.5 else y

reveal_type(g(f, g))
```

With stack trace

```
Error performing analysis: RangeError: Maximum call stack size exceeded
    at find (/usr/local/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/types.ts:2223:20)
    at Array.find (<anonymous>)
    at findSubtype (/usr/local/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/types.ts:2222:30)
    at isTypeSame (/usr/local/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/types.ts:2086:17)
    at canAssignType (/usr/local/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:19027:17)
    at canAssignFunction (/usr/local/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:20473:21)
    at canAssignType (/usr/local/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:19559:21)
    at canAssignFunctionParameter (/usr/local/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:19871:14)
    at canAssignFunction (/usr/local/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:20052:22)
    at ri (/usr/local/lib/node_modules/pyright/dist/pyright-internal/src/analyzer/typeEvaluator.ts:19559:21)
```

This is observed on pyright version 1.1.176 which might be quite stale.

**Comment by kmillikin** [comment]
I updated to 1.1.235 and now I observe that the type checker does not terminate (or takes a really long time, but not a stack overflow).

**Comment by erictraut** [comment]
This will be fixed in the next release. Apologies for taking so long on this issue. It was quite tricky to find the right fix, and it took me a couple of tries to get it right.

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.239, which I just published. It will also be included in the next release of pylance.
